File: elixir/hamming/hamming.exs
defmodule DNA do
  @doc """
  Returns number of differences between two strands of DNA, known as the Hamming Distance.

  ## Examples

  iex> DNA.hamming_distance('AAGTCATA', 'TAGCGATC')
  {:ok, 4}
  """
  @spec hamming_distance([char], [char]) :: non_neg_integer
  def hamming_distance(strand1, strand2) when length(strand1) != length(strand2) do
    {:error, "Lists must be the same length." }
  end

  @spec hamming_distance([char], [char]) :: non_neg_integer
  def hamming_distance(strand1, strand2) do
    {:ok, Enum.zip(strand1, strand2) |> Enum.count(fn {a,b} -> a != b end)}
  end
end

File: go/hamming/hamming.go
package hamming

import "errors"

func Distance(a, b string) (int, error) {
	if len(a) != len(b) {
		return -1, errors.New("strands must be of equal length")
	}

	distance := 0
	for idx := range a {
		if a[idx] != b[idx] {
			distance++
		}
	}

	return distance, nil
}
